qmxgraph/api.py

```diff
@@ -870,31 +869,40 @@ def is_cells_connectable(self):
     def run_layout(self, layout_name):
         return self.call_api('runLayout', layout_name)
 
-    def call_api(self, fn, *args):
+    def call_api(self, fn: str, *args, sync=True, check_api=True):
         """
         Call a function in underlying API provided by JavaScript graph.
 
-        :param str fn: A function call available in API.
-        :param Any args: Positional arguments passed to graph's
+        :param fn: A function call available in API.
+        :param args: Positional arguments passed to graph's
             JavaScript API call (unfortunately can't use named arguments
             with JavaScript). All object passed must be JSON encodable or
             Variable instances.
         :rtype: object
         :return: Return of API call.
         """
+        with self._call_context_manager_factory():
+            return self._call_api(fn, *args, sync=sync, check_api=check_api)
+
+    def _call_api(self, fn: str, *args, sync, check_api: bool=True):
         graph = self._graph()
         eval_js = graph.inner_web_view().eval_js
-        if qmxgraph.debug.is_qmxgraph_debug_enabled():
-            # Healthy check as if function didn't exist it just returns None,
-            # giving the impression that might have worked
-            if not graph.is_loaded():
-                raise qmxgraph.js.InvalidJavaScriptError(
-                    "Because graph is unloaded can't call the JavaScript API."
-                )
-            if eval_js("(typeof api === 'undefined') || !api.{}".format(fn)):
-                raise qmxgraph.js.InvalidJavaScriptError(
-                    'Unable to find function "{}" in QmxGraph '
-                    'JavaScript API'.format(fn))
+        # REMOVE extra checks: those were a good idea when we had the sync API, but now
```

Comment:
We could move this check into javascript to keep it sync (no extra blockers) with the actual call.

```diff
diff --git a/qmxgraph/api.py b/qmxgraph/api.py
index 6f4abcc..3b3ec76 100644
--- a/qmxgraph/api.py
+++ b/qmxgraph/api.py
@@ -887,22 +887,30 @@ class QmxGraphApi(object):
     def _call_api(self, fn: str, *args, sync, check_api: bool=True):
         graph = self._graph()
         eval_js = graph.inner_web_view().eval_js
-        # REMOVE extra checks: those were a good idea when we had the sync API, but now
-        # those calls are executing the event loop each time, which compounds errors.
-        # if sync and qmxgraph.debug.is_qmxgraph_debug_enabled() and check_api:
-        #     # Healthy check as if function didn't exist it just returns None,
-        #     # giving the impression that might have worked
-        #     if not graph.is_loaded():
-        #         raise qmxgraph.js.InvalidJavaScriptError(
-        #             "Because graph is unloaded can't call the JavaScript API."
-        #         )
-        #     if eval_js("(typeof api === 'undefined')"):
-        #         raise qmxgraph.js.InvalidJavaScriptError(
-        #             'API is undefined at this time')
-        #     if eval_js("!api.{}".format(fn)):
-        #         raise qmxgraph.js.InvalidJavaScriptError(
-        #             'Unable to find function "{}" in QmxGraph '
-        #             'JavaScript API'.format(fn))
-
-        call = qmxgraph.js.prepare_js_call(fn, *args)
-        return eval_js("api.{}".format(call), sync=sync, check_api=check_api)
+        call = f'api.{qmxgraph.js.prepare_js_call(fn, *args)}'
+
+        if qmxgraph.debug.is_qmxgraph_debug_enabled():
+            import textwrap
+            call = textwrap.dedent(
+                f'''\
+                try {{
+                    if (
+                        (typeof graphs === "undefined")
+                        || !graphs.isRunning()
+                    ) {{
+                        throw Error('`graphs` must be loaded and running');
+                    }}
+                    if (typeof api === "undefined") {{
+                        throw Error('`api` must be loaded')
+                    }}
+                    if (!api.{fn}) {{
+                        throw Error(
+                            'Unable to find function "{fn}"'
+                            + ' in QmxGraph JavaScript API'
+                        )
+                    }}
+                    {call};
+                }} finally {{ }}'''
+            )
+
+        return eval_js(call, sync=sync, check_api=check_api)
```

(github won't let me add this as a code change suggestion)

Reply:
Sounds good, created https://eden.esss.co/jira/browse/ASIM-4287 for it. 👍